189.configuration with implementation

meejah opened the pull request.

meejah commented:
(how is it possible that codecov is the first reported status? local tool says 91%)

meejah commented:
(Not sure what's up with the debian9 failures..)

meejah commented:
CodeCov is clearly just wrong: it is counting changes to `cli.py` around `_item_status` against this patch/project coverage but I haven't changed any of that code. Is it associating the wrong builds with the wrong PRs or something??

meejah commented:
(weird, again codecov is the first to complete..)

meejah commented:
come on, codecov: how did the "patch" coverage only hit 77% but somehow that also raised the project coverage to 87%??

meejah commented:
I marked this is as ready-for-review because we *could* merge it now -- or we could put arbitrarily more implementation in here (API, `magic-folder migrate`, tie into all the other commands, etc)

meejah commented:
I can't even tell what the "details" link in `codecov/patch` checker is trying to show me; it's got a whole bunch of lines that aren't touched by this PR in the visualization .. but maybe that's what it is supposed to do?

meejah commented:
See also some previous discussion here: https://github.com/LeastAuthority/magic-folder/pull/194

exarkun commented on docs/config.rst:
```diff
@@ -0,0 +1,180 @@
+found in a single directory. This directory can be specified via
+command-line option. (Should we have a default? Seems sensible, most
+OSes have a reasonably well-defined place for programs to store
+configuration data).
```
is there an API we can call, yet, that just returns the right path to use for all platforms? :/

meejah replied:
`#python` says https://pypi.org/project/appdirs/

exarkun commented on docs/config.rst:
```diff
@@ -0,0 +1,180 @@
+- "config"
+  - columns: "api_endpoint", "tahoe_client_url"
+  - `api_endpoint` is a Twisted server string description
+  - `tahoe_client_url` is a URL to our Tahoe-LAFS client's Web-UI
```
I guess the reason we might want the node directory instead of the http api root url is that some nodes are configured to listen on a random port  every startup?  And, specifically, gridsync configures its tahoe nodes to do this.  It might be annoying to have to update database state every time the data node restarts.

Or, we could mitigate this by *allowing* `tahoe_client_url` as a cli argument to `magic-folder run` and having it override the database value?

Or we could have a node directory in the db and read the node.url ourselves...

Or some combination of the above, I guess.

exarkun replied:
Pointing at the node directory means the magic-folder process could re-read a changed node.url itself and follow tahoe-lafs restarts/reconfigurations easily.

meejah replied:
okay, sounds good. also storing node-directory is definitely *safer* too (e.g. for anything else we forgot about that we need from tahoe still).

exarkun commented on docs/config.rst:
```diff
@@ -0,0 +1,180 @@
+  - column: "version"
+  - "version" is an int, currently `1`
+- "config" (will always contain 1 row)
+  - columns: "author_name", "author_private_key", "poll_interval"
```
This elemt is redundant I guess, since all the columns are listed below (and there are more than these three).

exarkun commented on docs/config.rst:
```diff
@@ -0,0 +1,180 @@
+  - columns: "author_name", "author_private_key", "poll_interval"
+  - "author_name" is a string of unicode
+  - "author_private_key" is a 32-byte blob (a NaCl Signing key)
+  - "stash_path" is a local path where we stash data awaiting upload
```
We could make this a path relative to the directory containing this database, making the whole config directory somewhat easily relocateable (or at least not imposing an additional barrier to relocation).

exarkun replied:
otoh since it involves rewriting configuration we probably want some kind of `magic-folder move-folder-state` command rather than just letting people `mv` stuff, so the path could always be rewritten by that tool.

exarkun commented on integration/test_general_cli.py:
```diff
@@ -1,5 +1,5 @@
 from os import mkdir
-from os.path import join
+from os.path import join, exists
```
Maybe we could gradually drift towards `twisted.python.filepath` and away from `os.path`.

exarkun commented on integration/test_general_cli.py:
```diff
@@ -44,3 +44,26 @@ def test_web_port_required(request, reactor, temp_dir, introducer_furl):
+    yield proto.done
+
+    assert exists(join(node_dir, "global.sqlite"))
+    assert exists(join(node_dir, "api_token"))
```
If this PR also adds some `magic-folder show-config` command that dumps the config (json?) then the test can assert that the output config matches the input config and remain a bit more oblivious to the exact internal representation.

exarkun replied:
Okay, good points.  `twist` has been a struggle all the way and if it's not even giving us what we want (eg ability to spit out clean json) maybe we should stop fighting with it and use something else.  Not right now, though, I suppose.  So perhaps we can just have a ticket to capture this and deal with it later.

exarkun commented on src/magic_folder/cli.py:
```diff
@@ -136,6 +144,57 @@
+
+        # validate
+        if exists(self['config']):
+            raise usage.UsageError("Directory '{}' already exists".format(self['config']))
```
Dang I wish anyone at all was working on twisted.python.usage.  These are such basic, common checks; it would be great if you could just declare them.

meejah replied:
There is always `Click` :)

exarkun commented on src/magic_folder/cli.py:
```diff
@@ -993,6 +1052,7 @@ class MagicFolderCommand(BaseOptions):
 
     subCommands = [
+        ["init", None, InitializeOptions, "Initialize a Magic Folder daemon."],
         ["create", None, CreateOptions, "Create a Magic Folder."],
```
I wonder if "init" vs "create" will be memorable for folks (whoever the folks that are directly interacting with this CLI are...).  I don't have any suggestions for different words to use, though.

meejah replied:
My secret plan is to re-think all the CLI commands .. but not necessarily all at once.

In some ways, maybe in makes sense for `init` to be `magic-folder create` and then there's a  `magic-folder add` or similar to create a new magic-folder inside the ... "magic folder daemon" (I don't have a better term yet) ... or we could just re-name `create` to `add` (now .. or later) and leave `init`. In any case, we probably need a common term for "magic-folder the whole daemon process" versus "magic-folder, the thing you can add many of to the daemon process".

exarkun commented on src/magic_folder/initialize.py:
```diff
@@ -0,0 +1,38 @@
+"""
+
+from twisted.python.filepath import (
+    FilePath,
```
:sunglasses:

exarkun commented on src/magic_folder/initialize.py:
```diff
@@ -0,0 +1,38 @@
+        tahoe_url,
+    )
+
+    return 0
```
`succeed(0)` perhaps?

exarkun commented on src/magic_folder/test/test_cli.py:
```diff
@@ -55,3 +67,20 @@ def test_good(self):
+
+    def tearDown(self):
+        super(TestInitialize, self).tearDown()
+        rmtree(self.temp.path, ignore_errors=True)
```
`FilePath.remove` is also a thing.  Additionally, `mktemp` creates the temporary directory beneath `_trial_temp` and trial deletes old `_trial_temp` directories when it starts up a new run.  This behavior makes me feel okay about tests leaving behind all of their temporary files.  They won't grow without bound and it's nice to automatically have them available for inspection instead of having to go out of your way to get them.

exarkun commented on src/magic_folder/test/test_config.py:
```diff
@@ -0,0 +1,166 @@
+
+    def setUp(self):
+        super(TestGlobalConfig, self).setUp()
+        self.temp = FilePath(self.mktemp())
```
A common thing to be broken is handling of non-ASCII paths.  It would be nice to exercise `create_global_configuration` with one of those.

exarkun commented on src/magic_folder/test/test_config.py:
```diff
@@ -0,0 +1,166 @@
+        rmtree(self.temp.path)
+
+    def test_create(self):
+        create_global_configuration(self.temp, "tcp:1234", "tcp:localhost:5678")
```
It's supposed to return a Deferred that succeeds with 0, maybe that's an okay assertion to make here?  Although it's a bit on the weak side.  Maybe this doesn't add much on top of the other happy-path tests - `test_load_db`, `test_rotate_api_key`, etc?

meejah replied:
This now tests different filenames (and `create_global_configuration` is synchronous).

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+(
+    name          TEXT PRIMARY KEY,  -- UTF8 name of this folder
+    location      TEXT               -- UTF8 path to this folder's configuration/state
+);
```
I *think* UTF8 bytes is the right choice here?  I'm not sure if there are some nasty surprises waiting on macOS or Windows though.  I think it wouldn't hurt to be explicit about these being bytes and not unicode (though "UTF8" pretty strongly implies bytes).  SQLite3 "type affinity" is a bit odd.  This is a Python 3 transcript but the basic idea holds for Python 2.  Notice the types of the 3 returned rows.

```
>>> import sqlite3
>>> conn = sqlite3.connect(":memory:")
>>> curs = conn.cursor()
>>> curs.execute("CREATE TABLE foo ( a TEXT )")
<sqlite3.Cursor object at 0x7f0b9f771d50>
>>> curs.execute("INSERT INTO foo VALUES (?)", (u"\N{SNOWMAN}",))
<sqlite3.Cursor object at 0x7f0b9f771d50>
>>> curs.execute("INSERT INTO foo VALUES (?)", (u"snowman",))
<sqlite3.Cursor object at 0x7f0b9f771d50>
>>> curs.execute("INSERT INTO foo VALUES (?)", (b"snowman",))
<sqlite3.Cursor object at 0x7f0b9f771d50>
>>> curs.execute("SELECT * FROM foo")
<sqlite3.Cursor object at 0x7f0b9f771d50>
>>> curs.fetchall()
[('☃',), ('snowman',), (b'snowman',)]
>>> 
```

meejah replied:
Mostly I'm looking at other SQLite3 code for inspiration; if there's something besides `TEXT` that we should use here (or some other comment / `.encode()` calls) I'm all ears :)

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+CREATE TABLE stash
+(
+    path TEXT    -- the path to our stash-directory
+);
```
Is this a left-over?  It seels like `config.stash_path` represents the same thing.

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+
+CREATE TABLE local_snapshots
+(
+    id VARCHAR(256) PRIMARY KEY, -- identifier (hash of .. stuff)
```
`VARCHAR(256)` is identical to `TEXT` in SQLite3.

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+        raise ValueError(
+            "'{}' already exists".format(basedir.path)
+        )
+    mkdir(basedir.path)
```
Race condition here.  `mkdir` and handle the EEXIST instead?  Also there's `basedir.createDirectory` and `basedir.makedirs`.

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+    mkdir(basedir.path)
+
+    # explain what is in this directory
+    with basedir.child("README").open("w") as f:
```
maybe mode `"wt"` and unicode strings for the text

meejah replied:
this was the source of hypothesis errors; i went with "wb" and a `.encode()` call ... but i'm not very confident that's correct.

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+        # XXX so, having the reactor here sucks. But if we pass in an
+        # IStreamServerEndpoint instead, how can we turn that back
+        # into an endpoint-string?
+        serverFromString(reactor, ep_string)
```
Yea, all that stuff you wrote.  There is no going from an IStreamServerEndpoint to a string, though.  The interface is not expressive enough.

And the endpoint parser is private so you can't really validate the endpoint in any way except with `serverFromString`.  That is pretty lame.  Maybe most of them don't use the reactor for parsing/initialization and we could get away with passing `None`? :/  Still pretty lame.

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+
+        mkdir(state_path.path)
+        stash_path = state_path.child("stash")
+        mkdir(stash_path.path)
```
It would be nice to (eventually) make this function atomic to simplify error recovery.  If we manage to make the state path and/or the stash path but then fail to populate the database schema, the magic-folder has not been created but you cannot try again with `create_magic_folder` without doing some manual cleanup of what should be internal state.

exarkun commented on src/magic_folder/config.py:
```diff
@@ -0,0 +1,419 @@
+            cursor.execute(
+                "INSERT INTO magic_folders VALUES (?, ?)",
+                (name, state_path.path)
+            )
```
This could fail if someone is doing something wacky - trying to create two magic-folders with the same name but different paths.  They'll race to this point and then one will succeed and the other will fail and leave intermediate state lying around.  Making the whole function atomic would probably address this.

meejah commented:
I believe I've addressed all the review comments. I didn't take any action on the migration code yet (that could be a follow-up? It seems the strategy there is to make all the table-creation code "migration code" so initial creation is just like "migrate from version 0 to version 1"?)

meejah commented:
There is now a `magic-folder migrate` (to exercise more of the code "for real") and a `magic-folder show-config` (so the integration test can have better asserts).

meejah commented:
(oops, missed adding a file)

meejah commented:
Local tool shows 91% coverage.
